src/app.js

```diff
@@ -61,90 +53,101 @@ export function app() {
-			let topic = '';
+		once(topic, handler) {
+			const wrappedHandler = message => {
+				handler(message);
+				this.off(topic, wrappedHandler);
+			};
+			this.on(topic, wrappedHandler);
+
 			/**
-			 * Data.
-			 * @optional
-			 * @type {*}
+			 * Return the deregistrator.
 			 */
-			let data = {};
-
-			if (arguments.length === 1 && typeof arguments[0] === 'object') {
-				/**
-				 * Single argument - {target, topic, data}
-				 */
-				target = arguments[0].target;
-				topic = arguments[0].topic;
-				data = arguments[0].data;
-			} else if (
-				arguments.length >= 2 &&
-				typeof arguments[0] === 'string' &&
-				typeof arguments[1] === 'string'
-			) {
-				/**
-				 * Two arguments - target, topic
-				 */
-				target = arguments[0];
-				topic = arguments[1];
-				if (arguments.length === 3) {
-					/**
-					 * Three arguments - target, topic, data
-					 */
-					data = arguments[2];
-				}
-			}
-			debug('%o (%o) to %o - %o', 'PUBLISH', topic, target, data);
-			postMessage({ type: 'PUBLISH', target, topic, data, token });
-			return this;
+			return () => this.off(topic, wrappedHandler);
 		},
 		/**
-		 * Request a response.
-		 * @returns {Promise}
+		 * Remove message handler.
+		 * @param  {string} topic Same as the topic which the handler uses.
+		 * @param  {Function} handler Reference to the same handler to delete.
+		 * @return {boolean} true on success.
 		 */
-		request: function() {},
+		off(topic, handler) {
+			const handlers = handlersByTopic.get(topic);
+			let deleted;
+			if (handlers) {
+				deleted = handlers.delete(handler);
+			}
+			debug(
+				'%s handler %o - %O',
+				deleted ? 'Deleted' : "Didn't find",
+				topic,
+				handler
+			);
+			return deleted;
+		},
 		/**
-		 * Spawn an app...
-		 * @alias spawn
-		 * @returns {Promise}
+		 * Publish message..
+		 * @param {string} target Target appId. - No wildcards supported
+		 * @param {string} topic Topic.
+		 * @param {*=} data Data.
 		 */
-		spawn: function() {},
+		emit(topic, target, data = {}) {
+			if (arguments.length < 2) {
+				throw new Error(
+					'ts.io().publish() called with invalid arguments.',
+					arguments
+				);
+			}
+			const message = { type: 'PUBLISH', target, topic, data, token };
+			if (token) {
+				debug('%o (%o) to %o - %o', 'PUBLISH', topic, target, data);
+				postMessage(message);
+			} else {
+				debug('%o (%o) to %o - %o', 'PUBLISH(queued)', topic, target, data);
+				queueMessage(message);
+			}
+		},
+		add(method, handler) {
+			methodHandlers.set(method, handler);
+		},
 		/**
-		 * Open an app..
-		 * @alias open
+		 * Do something RPC-style (spawn, request, etc.)
+		 * @async
+		 * @param {string} method Remote method to call (spawn, request, etc.)
+		 * @param {string} target Target appId. - No wildcards supported
+		 * @param {*=} data Data.
 		 * @returns {Promise}
 		 */
-		open: function() {}
+		async call(method, target, data = {}) {
+			switch (method) {
+				case 'spawn':
+					const message = {
```

Comment:
shouldn't `call` be generic? the spawn code doesn't look very tied to the `spawn` action

Reply:
I agree with this, `SPAWN` and `REQUEST` would work the same way, but then I guess we'll have to whitelist the possible commands, so people don't start misusing it instead of `EMIT`